src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
-    }
-#endif // DEBUG
+    private:
+        // Keep track of blocks with bbNum <= oldBlockMaxNum in a regular BlockSet, since
+        // it can hold all of them.
+        BlockSet oldBlocksInLoop; // Blocks with bbNum <= oldBlockMaxNum
 
-    flowList* pred;
-    flowList* predTop;
-    flowList* predEntry;
+        // Keep track of blocks with bbNum > oldBlockMaxNum in a separate BlockSet, but
+        // indexing them by (blockNum - oldBlockMaxNum); since we won't generate more than
+        // one new block per old block, this must be sufficient to track any new blocks.
+        BlockSet newBlocksInLoop; // Blocks with bbNum > oldBlockMaxNum
 
-    noway_assert(fgDomsComputed);
-    assert(fgHasLoops);
+        Compiler*    comp;
+        unsigned int oldBlockMaxNum;
 
-#if COUNT_LOOPS
-    hasMethodLoops         = false;
-    loopsThisMethod        = 0;
-    loopOverflowThisMethod = false;
-#endif
+    public:
+        LoopBlockSet(Compiler* comp) : comp(comp), oldBlockMaxNum(comp->fgBBNumMax)
+        {
+        }
 
-    /* We will use the following terminology:
-     * HEAD    - the basic block that flows into the loop ENTRY block (Currently MUST be lexically before entry).
-                 Not part of the looping of the loop.
-     * FIRST   - the lexically first basic block (in bbNext order) within this loop.  (May be part of a nested loop,
-     *           but not the outer loop. ???)
-     * TOP     - the target of the backward edge from BOTTOM. In most cases FIRST and TOP are the same.
-     * BOTTOM  - the lexically last block in the loop (i.e. the block from which we jump to the top)
-     * EXIT    - the loop exit or the block right after the bottom
-     * ENTRY   - the entry in the loop (not necessarly the TOP), but there must be only one entry
-     *
-     * We (currently) require the body of a loop to be a contiguous (in bbNext order) sequence of basic blocks.
-
-            |
-            v
-          head
-            |
-            |    top/beg <--+
-            |       |       |
-            |      ...      |
-            |       |       |
-            |       v       |
-            +---> entry     |
-                    |       |
-                   ...      |
-                    |       |
-                    v       |
-             +-- exit/tail  |
-             |      |       |
-             |     ...      |
-             |      |       |
-             |      v       |
-             |    bottom ---+
-             |
-             +------+
-                    |
-                    v
+        void reset(unsigned int seedBlockNum)
+        {
+            assert(seedBlockNum <= oldBlockMaxNum);
+            oldBlocksInLoop = BlockSetOps::MakeSingleton(comp, seedBlockNum);
+            newBlocksInLoop = BlockSetOps::MakeEmpty(comp);
+        }
 
-     */
+        bool isMember(unsigned int blockNum)
+        {
+            if (blockNum > oldBlockMaxNum)
+            {
```

Comment:
is there a reason you want it to be a `noway_assert` instead of a plain `assert`?